### PR TITLE
Don't try to reconnect client on interpreter shutdown

### LINKED ISSUE #6087: Error closing a local cluster when client still running
**What happened**:

When a local cluster (using processes) shuts down, a ton of errors are now spewed about scheduling new futures after shutdown.

I can't replicate it in my distributed dev environment, but in a different environment (which is quite similar, also running dask & distributed from main—just py py3.9.1 instead of py3.9.5?) the process hangs and never terminates until a ctrl-C. In my distributed dev environment, the same errors are spewed, but it exits (with code 0 no the less).

`git bisect` implicates https://github.com/dask/distributed/pull/6031 @graingert.

**Minimal Complete Verifiable Example**:

```python
# repro.py
import distributed
from distributed.deploy.local import LocalCluster


if __name__ == "__main__":
    cluster = LocalCluster(n_workers=1, threads_per_worker=1, processes=True)
    client = distributed.Client(cluster)
```

```python-traceback
(dask-distributed) gabe dev/distributed ‹f4c52e9a› » python repro.py
2022-04-07 15:47:02,414 - distributed.utils - ERROR - cannot schedule new futures after shutdown
Traceback (most recent call last):
  File "/Users/gabe/dev/distributed/distributed/comm/tcp.py", line 226, in read
    frames_nbytes = await stream.read_bytes(fmt_size)
tornado.iostream.StreamClosedError: Stream is closed

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/gabe/dev/distributed/distributed/client.py", line 1395, in _handle_report
    msgs = await self.scheduler_comm.comm.read()
  File "/Users/gabe/dev/distributed/distributed/comm/tcp.py", line 242, in read
    convert_stream_closed_error(self, e)
  File "/Users/gabe/dev/distributed/distributed/comm/tcp.py", line 150, in convert_stream_closed_error
    raise CommClosedError(f"in {obj}: {exc}") from exc
distributed.comm.core.CommClosedError: in <TCP (closed) Client->Scheduler local=tcp://127.0.0.1:58800 remote=tcp://127.0.0.1:58791>: Stream is closed

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/gabe/dev/distributed/distributed/utils.py", line 693, in log_errors
    yield
  File "/Users/gabe/dev/distributed/distributed/client.py", line 1225, in _reconnect
    await self._ensure_connected(timeout=timeout)
  File "/Users/gabe/dev/distributed/distributed/client.py", line 1254, in _ensure_connected
    comm = await connect(
  File "/Users/gabe/dev/distributed/distributed/comm/core.py", line 289, in connect
    comm = await asyncio.wait_for(
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/asyncio/tasks.py", line 481, in wait_for
    return fut.result()
  File "/Users/gabe/dev/distributed/distributed/comm/tcp.py", line 439, in connect
    stream = await self.client.connect(
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/site-packages/tornado/tcpclient.py", line 265, in connect
    addrinfo = await self.resolver.resolve(host, port, af)
  File "/Users/gabe/dev/distributed/distributed/comm/tcp.py", line 424, in resolve
    for fam, _, _, _, address in await asyncio.get_running_loop().getaddrinfo(
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/asyncio/base_events.py", line 856, in getaddrinfo
    return await self.run_in_executor(
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/asyncio/base_events.py", line 814, in run_in_executor
    executor.submit(func, *args), loop=self)
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/concurrent/futures/thread.py", line 161, in submit
    raise RuntimeError('cannot schedule new futures after shutdown')
RuntimeError: cannot schedule new futures after shutdown
2022-04-07 15:47:02,417 - distributed.utils - ERROR - cannot schedule new futures after shutdown
Traceback (most recent call last):
  File "/Users/gabe/dev/distributed/distributed/comm/tcp.py", line 226, in read
    frames_nbytes = await stream.read_bytes(fmt_size)
tornado.iostream.StreamClosedError: Stream is closed

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/gabe/dev/distributed/distributed/client.py", line 1395, in _handle_report
    msgs = await self.scheduler_comm.comm.read()
  File "/Users/gabe/dev/distributed/distributed/comm/tcp.py", line 242, in read
    convert_stream_closed_error(self, e)
  File "/Users/gabe/dev/distributed/distributed/comm/tcp.py", line 150, in convert_stream_closed_error
    raise CommClosedError(f"in {obj}: {exc}") from exc
distributed.comm.core.CommClosedError: in <TCP (closed) Client->Scheduler local=tcp://127.0.0.1:58800 remote=tcp://127.0.0.1:58791>: Stream is closed

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/gabe/dev/distributed/distributed/utils.py", line 693, in log_errors
    yield
  File "/Users/gabe/dev/distributed/distributed/client.py", line 1401, in _handle_report
    await self._reconnect()
  File "/Users/gabe/dev/distributed/distributed/client.py", line 1225, in _reconnect
    await self._ensure_connected(timeout=timeout)
  File "/Users/gabe/dev/distributed/distributed/client.py", line 1254, in _ensure_connected
    comm = await connect(
  File "/Users/gabe/dev/distributed/distributed/comm/core.py", line 289, in connect
    comm = await asyncio.wait_for(
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/asyncio/tasks.py", line 481, in wait_for
    return fut.result()
  File "/Users/gabe/dev/distributed/distributed/comm/tcp.py", line 439, in connect
    stream = await self.client.connect(
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/site-packages/tornado/tcpclient.py", line 265, in connect
    addrinfo = await self.resolver.resolve(host, port, af)
  File "/Users/gabe/dev/distributed/distributed/comm/tcp.py", line 424, in resolve
    for fam, _, _, _, address in await asyncio.get_running_loop().getaddrinfo(
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/asyncio/base_events.py", line 856, in getaddrinfo
    return await self.run_in_executor(
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/asyncio/base_events.py", line 814, in run_in_executor
    executor.submit(func, *args), loop=self)
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/concurrent/futures/thread.py", line 161, in submit
    raise RuntimeError('cannot schedule new futures after shutdown')
RuntimeError: cannot schedule new futures after shutdown
2022-04-07 15:47:02,471 - distributed.utils - ERROR - cannot schedule new futures after shutdown
Traceback (most recent call last):
  File "/Users/gabe/dev/distributed/distributed/comm/tcp.py", line 226, in read
    frames_nbytes = await stream.read_bytes(fmt_size)
tornado.iostream.StreamClosedError: Stream is closed

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/gabe/dev/distributed/distributed/client.py", line 1395, in _handle_report
    msgs = await self.scheduler_comm.comm.read()
  File "/Users/gabe/dev/distributed/distributed/comm/tcp.py", line 242, in read
    convert_stream_closed_error(self, e)
  File "/Users/gabe/dev/distributed/distributed/comm/tcp.py", line 150, in convert_stream_closed_error
    raise CommClosedError(f"in {obj}: {exc}") from exc
distributed.comm.core.CommClosedError: in <TCP (closed) Client->Scheduler local=tcp://127.0.0.1:58800 remote=tcp://127.0.0.1:58791>: Stream is closed

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/gabe/dev/distributed/distributed/utils.py", line 693, in log_errors
    yield
  File "/Users/gabe/dev/distributed/distributed/client.py", line 1521, in _close
    await asyncio.wait_for(asyncio.shield(handle_report_task), 0.1)
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/asyncio/tasks.py", line 481, in wait_for
    return fut.result()
  File "/Users/gabe/dev/distributed/distributed/client.py", line 1401, in _handle_report
    await self._reconnect()
  File "/Users/gabe/dev/distributed/distributed/client.py", line 1225, in _reconnect
    await self._ensure_connected(timeout=timeout)
  File "/Users/gabe/dev/distributed/distributed/client.py", line 1254, in _ensure_connected
    comm = await connect(
  File "/Users/gabe/dev/distributed/distributed/comm/core.py", line 289, in connect
    comm = await asyncio.wait_for(
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/asyncio/tasks.py", line 481, in wait_for
    return fut.result()
  File "/Users/gabe/dev/distributed/distributed/comm/tcp.py", line 439, in connect
    stream = await self.client.connect(
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/site-packages/tornado/tcpclient.py", line 265, in connect
    addrinfo = await self.resolver.resolve(host, port, af)
  File "/Users/gabe/dev/distributed/distributed/comm/tcp.py", line 424, in resolve
    for fam, _, _, _, address in await asyncio.get_running_loop().getaddrinfo(
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/asyncio/base_events.py", line 856, in getaddrinfo
    return await self.run_in_executor(
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/asyncio/base_events.py", line 814, in run_in_executor
    executor.submit(func, *args), loop=self)
  File "/Users/gabe/miniconda3/envs/dask-distributed/lib/python3.9/concurrent/futures/thread.py", line 161, in submit
    raise RuntimeError('cannot schedule new futures after shutdown')
```

**Environment**:

- Dask version: 6e307662be633695f57bce79458b77e9be2e5702
- Python version: 3.9.5
- Operating System: macOS
- Install method (conda, pip, source): source

### PR DESCRIPTION
**Pull request by mrocklin**
- [ ] Closes #6087
- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

**Review by crusaderky — CHANGES_REQUESTED** [review]
The import design is making me uneasy. It *works*, but if, in the future, an over-eager developer will move the import from inside the function to the top of the ``distributed.utils`` module, this will break and no tests will trip.
Could you move the ``atexit.register`` function from ``distributed/__init__.py`` to ``distributed/utils.py``? This way, any module that contains ``from distributed.utils import is_python_closing`` will guarantee that the atexit is registered well ahead of shutdown time.

Tested that this PR fixes the issue.

**Comment by mrocklin** [comment]
I like the idea, however, for some reason this change (which I think should do the job) fails to work.

```diff
diff --git a/distributed/__init__.py b/distributed/__init__.py
index 0d116516..e4a4be95 100644
--- a/distributed/__init__.py
+++ b/distributed/__init__.py
@@ -1,8 +1,6 @@
 from distributed import config  # isort:skip; load distributed configuration first
 from distributed import widgets  # isort:skip; load distributed widgets second
 
-import atexit
-
 import dask
 from dask.config import config  # type: ignore
 
@@ -74,18 +72,3 @@ def __getattr__(name):
         return __git_revision__
 
     raise AttributeError(f"module {__name__!r} has no attribute {name!r}")
-
-
-_python_shutting_down = False
-
-
-@atexit.register
-def _():
-    """Set a global when Python shuts down
-
-    See Also
-    --------
-    distributed.utils.is_python_shutting_down
-    """
-    global _python_shutting_down
-    _python_shutting_down = True
diff --git a/distributed/utils.py b/distributed/utils.py
index 24fab898..ce5f8ac3 100644
--- a/distributed/utils.py
+++ b/distributed/utils.py
@@ -1,6 +1,7 @@
 from __future__ import annotations
 
 import asyncio
+import atexit
 import contextvars
 import functools
 import importlib
@@ -1610,8 +1611,21 @@ def recursive_to_dict(
         tok.var.reset(tok)
 
 
+_python_shutting_down = False
+
+
 def is_python_shutting_down() -> bool:
     """Is the interpreter shutting down now?"""
-    from distributed import _python_shutting_down
-
     return _python_shutting_down
+
+
+@atexit.register
+def _():
+    """Set a global when Python shuts down
+
+    See Also
+    --------
+    distributed.utils.is_python_shutting_down
+    """
+    global _python_shutting_down
+    _python_shutting_down = True
```

`distributed.client` certainly imports from `distributed.utils`.  :thinking: .  Thoughts?

**Comment by mrocklin** [comment]
I'm feeling some time pressure to get this in soonish so that we have the option to release Friday.  

@crusaderky are your comments blockers, or can we resolve them later?

File: distributed/utils.py [hunk 1]
```diff
@@ -1608,3 +1608,10 @@ def recursive_to_dict(
         return repr(obj)
     finally:
         tok.var.reset(tok)
+
+
+def is_python_shutting_down() -> bool:
+    """Is the interpreter shutting down now?"""
+    from distributed import _python_shutting_down
```
Comment (fjetter) [inline distributed/utils.py:1615]: How about a compromise?
```suggestion
    # This import must remain local for the global variable to be
    # properly evaluated
    from distributed import _python_shutting_down
```

I don't want to block this but I'm wondering if this is not identical to https://docs.python.org/3/library/sys.html#sys.is_finalizing

Comment (crusaderky) [reply distributed/utils.py:1615]: I tried it and it doesn't work. It's worth adding a note in the docstring of the function though.

Comment (mrocklin) [reply distributed/utils.py:1615]: Yeah, same.  I was hopeful that it would.  That would have been very clean.

Comment (mrocklin) [reply distributed/utils.py:1615]: Happy to add comments, or do whatever really if anyone finds anything that works.  I'm not picky with this implementation.  I'm just trying to find a way to remove this issue.

Comment (mrocklin) [reply distributed/utils.py:1615]: That particular import isn't, I don't think, the issue.  I would be happy to move that to top level if it didn't cause a cycle.

The concern that I think @crusaderky was raising was that it's strange having this code in `__init__.py`.  Or at least that's my understanding.  Please correct me if I'm wrong.

Comment (crusaderky) [reply distributed/utils.py:1615]: If the function is registered at the end of ``distributed.__init__``, it works. 
If you keep the call to atexit.register in ``distributed.__init__``, but you move it just before ``from distributed.deploy import ...``, it breaks, as ``LocalCluster.__del__`` is invoked before the atexit callback.
I believe this is the same reason why ``sys.is_finalizing`` doesn't work.

This subtle behaviour should be explained in the docstring.

File: distributed/utils.py [hunk 2]
```diff
@@ -1608,3 +1608,10 @@ def recursive_to_dict(
         return repr(obj)
     finally:
         tok.var.reset(tok)
+
+
+def is_python_shutting_down() -> bool:
+    """Is the interpreter shutting down now?"""
```
Comment (crusaderky) [inline distributed/utils.py:1614]:
```suggestion
    """Is the interpreter shutting down now?
    This is a variant of ``sys.is_finalizing`` which can return True inside the ``__del__``
    method of classes defined inside the distributed package.
```

**Comment by crusaderky** [comment]
Happy to merge with the 3 suggestions above

File: distributed/utils.py [hunk 3]
```diff
@@ -1608,3 +1608,14 @@ def recursive_to_dict(
         return repr(obj)
     finally:
         tok.var.reset(tok)
+
+
+def is_python_shutting_down() -> bool:
+    """Is the interpreter shutting down now?
+    This is a variant of ``sys.is_finalizing`` which can return True inside the ``__del__``
+    method of classes defined inside the distributed package.
```
Comment (jrbourbeau) [inline distributed/utils.py:1616]:
```suggestion
    method of classes defined inside the distributed package.
    """
```

Comment (crusaderky) [reply distributed/utils.py:1617]: ? the docstring is already closed below

**Review by crusaderky — APPROVED** [review]

File: distributed/utils.py (hunk 3 above)
Comment (mrocklin) [reply distributed/utils.py:1617]: It wasn't before.  There was a bad merge from a github suggestion (my mistake).  James was correct to point it out.  I fixed it in a commit in between your two comments here. 

Anyway, all is well.

**Comment by mrocklin** [comment]
I saw a couple of unrelated intermittent failures on this PR.  I don't think that this is likely to cause any issues, but it's strange enough that I'm rerunning everything for good measure to see if it's a pattern.

**Pull request merged**